### PR TITLE
refactor(training): make ProfilingConfig authoritative

### PR DESCRIPTION
**Pull request by cuichenx**
## Summary

Make `ProfilingConfig` authoritative across supported training/inference/tool entrypoints, profiler controls, memory history, and checkpoint output.

- Normalize CLI/YAML inputs once; pass profiling configuration explicitly to consumers.
- Keep runtime profiler handles separate from serialized settings and pass iteration progress explicitly.
- Preserve `--profile` gating, profiler modes, ranks, windows, NVTX, and memory-snapshot behavior.
- Save config-owned fields through a detached legacy args snapshot without changing live args. Resume retains the current run's profiling policy, matching existing checkpoint precedence.

No new config registry, args fallback mode, checkpoint format, or dependency on #7344/authoritative TrainState.

## Base and scope

Replaces #7564, which closed when its temporary base branch was deleted after #7418 merged. Targets `main`, incorporating `5e85e2b27fc29bcf938e8f3b38acfbdecffcc8f3` through a normal merge; published history is preserved.

Removed unrelated import sorting, whitespace cleanup, and unnecessary reformatting. The remaining 50 files implement profiling ownership or update required callers/tests; no formatting-only files. Upstream context-parallel checkpoint and packed-sequence logging changes are retained.

## Validation

Replacement-head checks: all 50 changed Python files parse, imported/direct/partial-call audit found no missing profiling arguments, and the diff passes whitespace checks. Fresh independent full-diff review is clean. GPU/runtime validation has **not** been rerun after merging the new main base.

Prior-revision evidence only: tested `dc57fd4d85e4980d057945307c3c0e88cf96470a` in the same NeMo 26.10.rc0 environment on both sides:

- Targeted GPU unit suite: 184/184 candidate tests and 158/158 parent tests pass, including 26 profiling ownership cases and actual checkpoint serialization in three formats.
- Eight short runtime comparisons pass exactly against the #7418 parent and separately against pinned main `3226303b34002c6022c08035694da0b46e4110b1`: fresh/save, resume, disabled profiling, excluded rank with an unused invalid window, nsys/NVTX, PyTorch profiling, memory history, and resume profiling policy.
- 13 fork-only harness self-tests pass in each allocation. PyTorch traces and memory snapshots were produced on both sides. No configuration differences, execution/capture errors, or new test failures.
- Changed-file Black/pylint/isort and copyright checks passed on that prior revision.

Reproduction and limits: [fork-only test suite](https://github.com/cuichenx/Megatron-LM/tree/chcui/omar/config-seam-tests/tools/config_seam_tests) and [profiling results](https://github.com/cuichenx/Megatron-LM/blob/chcui/omar/config-seam-tests/tools/config_seam_tests/PROFILING_RESULTS.md). This is targeted equivalence and short-training validation, not a full-suite, convergence, or performance claim.

The replacement's required repository-wide pre-commit check found 26 existing formatting files and unrelated Pylint findings; isort passed. Formatter-generated changes were discarded, leaving the PR focused on ownership changes. No GPU results above are attributed to the new merged head.

### CODE REVIEW

**Comment by cuichenx** [comment]
Validated `098a809c25df41907fca9eca9aa7c19eee6b27b2` against frozen main `5e85e2b27fc29bcf938e8f3b38acfbdecffcc8f3` in NeMo 26.10.rc0:

- Targeted GPU units: 325 candidate / 294 baseline passed, with the same three skips; includes 31 profiling/global-config cases.
- [Fork-only seam suite](https://github.com/cuichenx/Megatron-LM/tree/chcui/omar/config-seam-tests/tools/config_seam_tests): all eight runtime comparisons pass exactly, with zero captured differences or execution errors. All 17 harness self-tests pass; both revisions produced profiler traces and memory snapshots.
- Independent complete-diff review and changed-file checks pass. Existing repository-wide formatting/Pylint findings remain separate.

[Current results and pinned revisions](https://github.com/cuichenx/Megatron-LM/blob/chcui/omar/config-seam-tests/tools/config_seam_tests/PROFILING_RESULTS.md). Normal push; no history rewrite. The revision removes TrainingProfiler/new profiling parameters and uses #7344's global config API.

**Comment by cuichenx** [comment]
Updated in `edc8f67b2ae1e8f1469267a904d1208b6c250bfd`: consumers acquire `cfg = get_run_config()` and read `cfg.profiling.x`. No signature/import/behavior changes or unrelated formatting.

Targeted GPU unit suite rerun in NeMo 26.10.rc0: **325 candidate / 294 frozen-baseline passed, three identical skips**, including all 31 profiling/global-config cases. Baseline: `5e85e2b27fc29bcf938e8f3b38acfbdecffcc8f3`.

The eight-case runtime parity result from the [fork-only suite](https://github.com/cuichenx/Megatron-LM/tree/chcui/omar/config-seam-tests/tools/config_seam_tests) remains evidence for `098a809c25df41907fca9eca9aa7c19eee6b27b2`, not a new runtime run on this head. This follow-up is AST-equivalent after reversing only the requested access-path substitution.

**Comment by cuichenx** [comment]
Full original **38-case suite passed** on `edc8f67b2ae1e8f1469267a904d1208b6c250bfd` versus frozen main `5e85e2b27fc29bcf938e8f3b38acfbdecffcc8f3`, using NeMo 26.10.rc0:

- 37 exact passes: 27 builder checks and ten short runtime checks across 1/2/8 H100s, including save/resume, recompute, fine-tuning, frozen-vision VLM, EP/CP/TP/PP/VPP/DP coverage.
- One expected CLI rejection: `invalid_batch_conflict` rejects conflicting batch-size flags on both sides.
- Zero captured differences or execution/capture errors; 17 harness self-tests pass in every allocation.

[Fork-only suite and reproduction instructions](https://github.com/cuichenx/Megatron-LM/tree/chcui/omar/config-seam-tests/tools/config_seam_tests) · [Detailed results](https://github.com/cuichenx/Megatron-LM/blob/chcui/omar/config-seam-tests/tools/config_seam_tests/PROFILING_RESULTS.md)

Follow-up `4d3adb4cbb0290cc71a0fb012408dcd97223a40d` only fixes test-import formatting/order; production code is unchanged and GitHub linting is green. The matrix was run on `edc8f67b2`, not relabeled as a run on the lint-only commit. This establishes captured-config equivalence and short-runtime completion, not convergence or performance parity.

**Comment by cuichenx** [comment]
Merged main `ffb3c4f966f000945578ae434dd85b76167677ea` normally and resolved the #7344 overlap. Latest tested/pushed head: `b57293693a1ac6fe0c6215ec746e9ee6acd7fa41`.

Fixed the CI failures: both parallel GDN tests now request the run-config fixture. Updated that fixture to a serializable container and supplied profiling defaults in #7344's checkpoint-YAML test mock. The upstream registry and single registration remain unchanged.

Validation in NeMo 26.10.rc0:

- **220 focused unit tests passed**, including all five checkpoint-YAML tests.
- **3/3 exact before/after runtime comparisons passed**: fresh/save, PyTorch profiling, profiling after resume ([fork-only suite](https://github.com/cuichenx/Megatron-LM/tree/chcui/omar/config-seam-tests/tools/config_seam_tests)).
- Changed-file lint and independent review passed.

The local four-GPU GDN test gets past the original config-initialization failure, but hits `input_grad_baseline contains nan`; the identical test reproduces that failure on frozen main. This baseline/environment issue stays separate; full parallel-GDN and CI green status are not claimed. The original 38-case suite was not rerun on this merge. CODEOWNERS approval remains a human-review gate.

File: megatron/training/training.py [hunk 1]
```diff
@@ -3636,6 +3637,7 @@ def training_log(
     """Log training information such as losses, timing, ...."""
     callback_manager = normalize_callbacks(callback_manager)
     args = get_args()
```
Comment (cuichenx) [inline megatron/training/training.py]: add comment explaining the transition, duplicating during transition

File: megatron/training/training.py [hunk 2]
```diff
@@ -3636,6 +3641,9 @@ def training_log(
...
+    # Temporary args/config duplication during the training-loop refactor:
+    # profiling is config-owned; unmigrated consumers still use legacy args.
```
Comment (cuichenx) [inline megatron/training/training.py]: don't mention profiling in this comment, since this is the first PR out of many config migrations

**Comment by cuichenx** [comment]
Removed `checkpoint_args_snapshot`; checkpoint production code now matches frozen main. Legacy checkpoint args remain unchanged, and `run_config.yaml` retains the authoritative config.

Retested **`bca9487706c4062b654c4c1317d61c212bacd0dc`** against frozen main `ffb3c4f966f000945578ae434dd85b76167677ea` with NeMo 26.10.rc0:

- 219 targeted unit tests passed (profiling ownership, initialization, argument construction, checkpointing, and upstream run-config YAML tests).
- 3/3 exact runtime comparisons passed: `runtime_fresh`, `profiling_pytorch`, `profiling_resume`.
- Independent review and changed-file lint checks passed; unrelated all-files baseline lint findings remain separate.

Used the [fork-only seam suite](https://github.com/cuichenx/Megatron-LM/tree/chcui/omar/config-seam-tests/tools/config_seam_tests) with the local 110-case draft capture extensions (22 harness self-tests passed). Only the three runtime cases above were run for this head; this is **not** a full 38/110-case result.
